src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,120 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_LENGTH = 10;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        this.elementData = (T[]) new Object[DEFAULT_LENGTH];
+        size = 0;
+    }
 
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheck(index);
+        if (size == elementData.length) {
+            grow();
+        }
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        T[] listToAdd = (T[]) new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            listToAdd[i] = list.get(i);
+        }
+        System.arraycopy(listToAdd, 0, elementData, size, listToAdd.length);
+        size += listToAdd.length;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheck(index);
+        T result = elementData[index];
+        System.arraycopy(elementData, index + 1,
+                elementData, index, size - index);
+        size -= 1;
+        return result;
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        if (t == null) {
+            for (int i = 0; i < size; i++) {
+                if (elementData[i] == null) {
+                    return remove(i);
+                }
+            }
+        } else {
+            for (int i = 0; i < size; i++) {
+                if (t.equals(elementData[i])) {
+                    return remove(i);
+                }
+            }
+        }
```

Comment:
simplify this to one loop with combined condition

Reply:
simplified to one loop